### DISCUSSION TITLE
Is there a way to combine multiple highlighters in console?

### DISCUSSION THREAD

**Discussion by ahuang11** [Textualize/rich#2460] [Q&A] [answered]
I would like to pass a list to highlighter keyword. Is it possible?
```
from rich.console import Console
from rich.highlighter import RegexHighlighter, JSONHighlighter
from rich.theme import Theme

class EmailHighlighter(RegexHighlighter):
    """Apply style to anything that looks like an email."""

    base_style = "example."
    highlights = [r"(?P<email>[\w-]+@([\w-]+\.)+[\w-]+)"]


theme = Theme({"example.email": "bold magenta", "json.number": "red"})
console = Console(highlighter=[EmailHighlighter(), JSONHighlighter()], theme=theme)
console.print("Send funds to money@example.org")
```

Or would I have to combine the highlighters with a custom Highlighter?
```
class CustomHighlighter(RegexHighlighter):
    base_style = "custom."
    highlights = EmailHighlighter.highlights + JSONHighlighter.highlights
```

**Comment by willmcgugan** [comment, marked as answer]
A list won't work. But a highlighter that combines both should work.

**Reply by ahuang11** [reply]
Thanks for the response!

To clarify, if unable to specify multiple highlighters, then all the base_style will be "custom" like this?
`theme = Theme({"custom.email": "bold magenta", "custom.number": "red"})`

I'm wondering what the base style is used for in this case.

**Comment by maltevesper** [comment]
I ran into the same issue, and while trying to create an issue I found this thread. I think documenting this better could help avoid questions, or feature requests of the form "please add X to the default highlighter". Here are a few things I think could help a few people:

- Add an example of extending the default highlighter to the documentation: i.e. 
  ```
  class HashHighlighter(ReprHighlighter):
      """Apply style to anything that looks like an email."""

      def __init__(self) -> None:
          super().__init__()
          self.__class__.highlights.append(
              r"\b(?P<git_hash>(?=[a-fA-F]+[0-9]+|[0-9]+[a-fA-F]+)[a-fA-F0-9]{7}|[a-fA-F0-9]{40})\b"
          )
  console = Console(highlighter=HashHighlighter())
  console.push_theme(Theme({"repr.git_hash": "bold magenta"}))
  ```
- Consider adding another way to stack highlighters, i.e. How about providing a highlighter, which calls multiple highlighters?
  `StackedHighlighter([ReprHighlighter,CustomHighlighter])` 
  this would avoid having the theme issue, that new regexes also need theme extensions.

I would be willing to chip in on a PR, but given that the project has a bunch of slightly dated open PRs I do not want to do so, unless I know there is a chance for the PR to be looked at.

(Note: the init thing looksd ugly to me, for a reason I have not figured out yet the following did not work:
```
  class HashHighlighter(ReprHighlighter):
     highlights = [
         *ReprHighlighter.highlights,
         r"\b(?P<git_hash>(?=[a-fA-F]+[0-9]+|[0-9]+[a-fA-F]+)[a-fA-F0-9]{7}|[a-fA-F0-9]{40})\b",
     ]
```

**Reply by maltevesper** [reply]
Oh, and on a side note, I wonder if performance wise it would make sense to `re.compile` the regexes in RegexHighlighter on init, rather than every time we use it to highlight something. Again: I am willing to look into this (create a MR and profile it), IFF this is something the project would potentially merge.
